test/src/unit-conversions.cpp

```diff
@@ -1710,7 +1710,8 @@ TEST_CASE("std::optional")
         std::optional<std::string> opt_null;
 
         CHECK(json(opt_null) == j_null);
-        CHECK(std::optional<std::string>(j_null) == std::nullopt);
+        CHECK_THROWS_WITH(std::optional<std::string>(j_null) == std::nullopt,
```

Comment:
Why is `null` not converted to a `nullopt` here?

Reply:
See 2.b in my comments.
> I agree with this view:
>> The reason is the template optional(U&&) constructor template, which is supposed to activate when T (int) is constructible from U and U is neither std::in_place_t nor optional, and direct-initialize T from it. And so it does, stamping out optional(foo&).

`j_null` is basic_json, and there is a way that basic_json can be converted to `string` by calling `void from_json(const BasicJsonType& j, typename BasicJsonType::string_t& s)`

Reply:
This may be a technical explanation, but as I user, when I ask to translate a JSON value to a `std::optional<std::string>`, then I expect this behavior:

- string values are stored as value,
- `null` is stored as `std::nullopt`, and
- all other values throw

At least this is what the code

```cpp
#ifdef JSON_HAS_CPP_17
template<typename BasicJsonType, typename T>
void from_json(const BasicJsonType& j, std::optional<T>& opt)
{
    if (j.is_null())
    {
        opt = std::nullopt;
    }
    else
    {
        opt = j.template get<T>();
    }
}
#endif
```

tries to accomplish.